File: packages/tracelet_sync/android/src/main/kotlin/com/ikolvi/tracelet_sync/TraceletSyncPlugin.kt
package com.ikolvi.tracelet_sync

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import com.ikolvi.tracelet.sdk.TraceletSdk
import com.ikolvi.tracelet.sdk.location.LocationDataSink
import com.ikolvi.tracelet.sdk.sync.NO_SYNC_BODY_BUILDER_SENTINEL
import uniffi.tracelet_sync.SyncManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class TraceletSyncSink(private val sdk: TraceletSdk) : LocationDataSink, TraceletSdk.SyncProvider {
    // SupervisorJob: a single failed sync must not cancel the scope, else the
    // first background sync that throws kills every future sync (Issue #134).
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val syncMutex = Mutex()
    private val syncManager = SyncManager()

    private var syncJob: kotlinx.coroutines.Job? = null
    private val DEBOUNCE_MS = 10_000L

    override fun insertLocation(location: Map<String, Any?>) {
        val delayMs = sdk.rustEngineState?.getConfig()?.http?.autoSyncDelay?.toLong() ?: 10_000L
        if (syncJob?.isActive == true) return
        syncJob = scope.launch {
            // Contain any throwable so a single failed iteration can't tear down
            // auto-sync; re-throw real cancellation so stop() still works (#134).
            try {
                kotlinx.coroutines.delay(delayMs)
                triggerSync()
            } catch (ce: kotlinx.coroutines.CancellationException) {
                throw ce
            } catch (t: Throwable) {
                sdk.logger.error("TraceletSyncSink: auto-sync iteration failed (contained): ${t.message}")
            }
        }
    }

    // Cancel the in-flight debounce so stop() takes effect immediately (#213).
    // TraceletSdk.stop() calls syncProvider?.cancelPendingSync(); without this
    // override (the interface default is a no-op) the coroutine kept sleeping
    // through delay(autoSyncDelay) and still fired an HTTP sync up to ~10s after
    // the user stopped tracking.
    override fun cancelPendingSync() {
        syncJob?.cancel()
        syncJob = null
        sdk.logger.debug("TraceletSyncSink: pending debounced background sync cancelled on stop().")
    }
    
    private suspend fun triggerSync() {
        syncMutex.withLock {
            sdk.logger.debug("triggerSync started")
            val db = sdk.rustDatabase ?: run {
                sdk.logger.error("rustDatabase is null")
                return
            }
            val state = sdk.rustEngineState ?: run {
                sdk.logger.error("rustEngineState is null")
                return
            }
            
            try {
                val coreHttp = state.getConfig().http
                sdk.logger.debug("coreHttp config: url=${coreHttp.url}, autoSync=${coreHttp.autoSync}")
                if (coreHttp.url.isNullOrEmpty() || !coreHttp.autoSync) return
                
                val limit = if (coreHttp.maxBatchSize > 0) coreHttp.maxBatchSize else 250
                val records = db.getLocationsBatch(uniffi.tracelet_core.LocationQuery(
                    startTimeMs = null,
                    endTimeMs = null,
                    limit = limit.toInt(),
                    offset = null,
                    // Honor the configured sort order (0=ascending, 1=descending)
                    // instead of always defaulting to ascending (Issue #138).
                    orderDescending = coreHttp.locationsOrderDirection == 1
                ))
                sdk.logger.debug("Found ${records.size} locations in DB")
                if (records.isEmpty()) return
                
                val count = syncBatchBlocking(coreHttp, records)
                if (count > 0L) {
                    records.lastOrNull()?.id?.let { lastId ->
                        db.clearLocationsUpTo(lastId)
                        sdk.logger.info("Synced and cleared $count locations.")
                    }
                }
            } catch (e: Exception) {
                sdk.logger.error("Sync failed: ${e.message}")
            }
        }
    }
    
    override fun syncBatchBlocking(config: uniffi.tracelet_core.HttpConfig, records: List<uniffi.tracelet_core.DbLocationRecord>): Long {
        val syncConfig = uniffi.tracelet_sync.SyncHttpConfig(
            url = config.url,
            method = config.method,
            headers = config.headers,
            batchSync = config.batchSync,
            maxBatchSize = config.maxBatchSize,
            autoSync = config.autoSync,
            maxRetries = config.maxRetries,
            retryBackoffBase = config.retryBackoffBase,
            retryBackoffCap = config.retryBackoffCap,
            sslPinningCertificates = config.sslPinningCertificates,
            sslPinningFingerprints = config.sslPinningFingerprints,
            httpRootProperty = config.httpRootProperty,
            params = config.params,
            extras = config.extras,
            disableAutoSyncOnCellular = config.disableAutoSyncOnCellular,
            enableDeltaCompression = config.enableDeltaCompression,
            deltaCoordinatePrecision = config.deltaCoordinatePrecision,
            locationsOrderDirection = config.locationsOrderDirection
        )
        val syncRecords = records.map {
            uniffi.tracelet_sync.SyncLocationRecord(
                id = it.id,
                uuid = it.uuid,
                timestamp = it.timestamp,
                latitude = it.latitude,
                longitude = it.longitude,
                accuracy = it.accuracy,
                speed = it.speed,
                heading = it.heading,
                altitude = it.altitude,
                isMock = it.isMock,
                isMoving = it.isMoving,
                activity = it.activity,
                event = it.eventType,
                routeContext = it.routeContext,
                address = it.address  // #212: carry reverse-geocoded address into the default payload
            )
        }
        val interceptor = sdk.dartSyncInterceptor
        sdk.logger.debug("TraceletSyncPlugin Interceptor is $interceptor")
        if (interceptor != null) {
            // Issue #126: emit the SAME nested schema as onLocation/getLocations
            // (nested coords/activity/battery + route_context) so the Dart
            // custom-body builder receives a consistent shape instead of a flat
            // map with a raw String activity.
            val recordMaps = records.map { sdk.mapRecordToLocation(it) }
            sdk.logger.debug("Calling requestSyncBody on interceptor with ${recordMaps.size} records")
            val customBody = interceptor.requestSyncBody(recordMaps)
            sdk.logger.debug("requestSyncBody returned: $customBody")
            if (customBody == null) {
                // Builder registered but failed → abort (0 = nothing synced).
                sdk.logger.error("Custom sync body failed to build; aborting sync")
                sdk.getEventSender().sendHttp(mapOf(
                    "success" to false,
                    "status" to 0,
                    "responseText" to "custom sync body failed to build",
                    "isRetry" to false,
                    "retryCount" to 0
                ))
                return 0L
            }
            if (customBody != NO_SYNC_BODY_BUILDER_SENTINEL) {
                return kotlinx.coroutines.runBlocking {
                    val success = executeFallbackHttpSync(config, customBody, interceptor)
                    sdk.logger.debug("executeFallbackHttpSync success: $success")
                    if (success) {
                        sdk.getEventSender().sendHttp(mapOf(
                            "success" to true,
                            "status" to 200,
                            "responseText" to "Synced ${records.size} locations via custom body",
                            "isRetry" to false,
                            "retryCount" to 0
                        ))
                        records.size.toLong()
                    } else {
                        sdk.getEventSender().sendHttp(mapOf(
                            "success" to false,
                            "status" to 0,
                            "responseText" to "Custom body sync failed",
                            "isRetry" to false,
                            "retryCount" to 0
                        ))
                        0L
                    }
                }
            }
            // sentinel → no builder → fall through to the default sync below.
        }

        try {
            val count = syncManager.syncBatchBlocking(syncConfig, syncRecords).toLong()
            if (count > 0L) {
                sdk.getEventSender().sendHttp(mapOf(
                    "success" to true,
                    "status" to 200,
                    "responseText" to "Synced $count locations",
                    "isRetry" to false,
                    "retryCount" to 0
                ))
            } else {
                sdk.getEventSender().sendHttp(mapOf(
                    "success" to false,
                    "status" to 0,
                    "responseText" to "Sync failed",
                    "isRetry" to false,
                    "retryCount" to 0
                ))
            }
            return count
        } catch (e: Exception) {
            sdk.logger.error("Sync failed: ${e.message}")
            sdk.getEventSender().sendHttp(mapOf(
                "success" to false,
                "status" to 0,
                "responseText" to (e.message ?: "Unknown error"),
                "isRetry" to false,
                "retryCount" to 0
            ))
            throw e
        }
    }

    private suspend fun executeFallbackHttpSync(
        coreHttp: uniffi.tracelet_core.HttpConfig,
        customBody: String,
        interceptor: com.ikolvi.tracelet.sdk.sync.DartSyncInterceptor?
    ): Boolean {
        var currentHeaders = coreHttp.headers
        val maxRetries = coreHttp.maxRetries.toInt()
        
        for (attempt in 0..maxRetries) {
            try {
                val url = java.net.URL(coreHttp.url)
                val conn = url.openConnection() as java.net.HttpURLConnection
                conn.requestMethod = if (coreHttp.method.toInt() == 1) "PUT" else "POST"
                conn.doOutput = true
                conn.setRequestProperty("Content-Type", "application/json")
                conn.connectTimeout = 15000
                conn.readTimeout = 15000

                if (currentHeaders.isNotEmpty()) {
                    try {
                        currentHeaders.forEach { (key, value) ->
                            conn.setRequestProperty(key, value)
                        }
                    } catch (e: Exception) {
                        sdk.logger.error("Failed to set HTTP headers: ${e.message}")
                    }
                }

                conn.outputStream.use { os ->
                    val input = customBody.toByteArray(kotlin.text.Charsets.UTF_8)
                    os.write(input, 0, input.size)
                }

                val status = conn.responseCode
                conn.disconnect()
                
                if (status in 200..299) {
                    return true
                } else if (status == 401 && interceptor != null) {
                    if (interceptor.requestTokenRefresh()) {
                        val newConfig = sdk.rustEngineState?.getConfig()?.http
                        if (newConfig != null) {
                            currentHeaders = newConfig.headers
                        }
                        continue
                    }
                }
            } catch (e: Exception) {
                sdk.logger.error("HTTP Sync failed: ${e.message}")
            }
            if (attempt < maxRetries) {
                kotlinx.coroutines.delay(1000L * (attempt + 1))
            }
        }
        return false
    }
}

/** TraceletSyncPlugin */
class TraceletSyncPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private var syncSink: TraceletSyncSink? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "tracelet_sync")
        channel.setMethodCallHandler(this)
        
        try {
            val context = binding.applicationContext
            val traceletSdk = TraceletSdk.getInstance(context)
            
            val sink = TraceletSyncSink(traceletSdk)
            traceletSdk.registerSyncProvider(sink)
            syncSink = sink
            
            traceletSdk.logger.info("Sync sink registered!")
        } catch (e: Exception) {
            val ctx = binding.applicationContext
            TraceletSdk.getInstance(ctx).logger.error("Failed to init sync engine: ${e.message}")
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "initialize") {
            result.success(true)
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: packages/tracelet_sync/ios/tracelet_sync/Sources/tracelet_sync/TraceletSyncPlugin.swift
import Flutter
import UIKit
import TraceletSDK

actor SyncCoordinator {
    private var isSyncing = false
    private var syncTask: Task<Void, Never>?
    
    func scheduleSync(sink: TraceletSyncSink) {
        let delayMs = TraceletSdk.shared.rustEngineState?.getConfig().http.autoSyncDelay ?? 10000
        let delayNanos = UInt64(delayMs) * 1_000_000
        
        if syncTask != nil { return }
        
        // CRITICAL FIX: Explicitly request background execution time.
        // Without this, iOS will suspend the app immediately after the location event completes,
        // preventing the HTTP sync task from ever firing.
        let bgTaskId = BackgroundTaskHelper.shared.begin("tracelet_sync_task")
        
        syncTask = Task {
            try? await Task.sleep(nanoseconds: delayNanos)
            // Bail if the debounce was cancelled while sleeping (stop(), #213).
            if Task.isCancelled {
                self.syncTask = nil
                BackgroundTaskHelper.shared.end(bgTaskId)
                return
            }
            self.syncTask = nil
            await self.triggerSync(sink: sink)
            BackgroundTaskHelper.shared.end(bgTaskId)
        }
    }

    /// Cancel the in-flight debounce so stop() halts background sync immediately (#213).
    func cancel() {
        syncTask?.cancel()
        syncTask = nil
    }

    func triggerSync(sink: TraceletSyncSink) async {
        guard !isSyncing else { return }
        isSyncing = true
        defer { isSyncing = false }
        
        guard let db = TraceletSdk.shared.rustDatabase,
              let state = TraceletSdk.shared.rustEngineState else {
            return
        }
        
        do {
            let updatedHttp = state.getConfig().http
            guard let url = updatedHttp.url, !url.isEmpty else { return }
            guard updatedHttp.autoSync else { return }
            
            TraceletSdk.shared.logger.debug("Triggering sync to URL: \(url)")
            
            let limit: Int32 = updatedHttp.maxBatchSize > 0 ? updatedHttp.maxBatchSize : 250
            let coreRecords = try db.getLocationsBatch(query: LocationQuery(
                startTimeMs: nil,
                endTimeMs: nil,
                limit: limit,
                offset: nil,
                orderDescending: nil
            ))
            TraceletSdk.shared.logger.debug("Found \(coreRecords.count) locations in DB.")
            if coreRecords.isEmpty { return }
            
            // CRITICAL FIX: Offload the synchronous blocking FFI call to a background 
            // DispatchQueue so we do not starve the Swift Concurrency cooperative thread pool.
            let count = try await withCheckedThrowingContinuation { continuation in
                DispatchQueue.global(qos: .utility).async {
                    do {
                        let result = try sink.syncBatchBlocking(config: updatedHttp, records: coreRecords)
                        continuation.resume(returning: result)
                    } catch {
                        continuation.resume(throwing: error)
                    }
                }
            }
            
            if count > 0, let lastId = coreRecords.last?.id {
                try db.clearLocationsUpTo(maxId: lastId)
                TraceletSdk.shared.logger.debug("Synced and cleared \(count) locations.")
            } else {
                TraceletSdk.shared.logger.debug("No locations synced or count was 0.")
            }
        } catch {
            TraceletSdk.shared.logger.debug("Sync failed with error: \(error)")
        }
    }

struct FallbackSyncResult {
    let success: Bool
    let status: Int
    let responseText: String
}

    // `nonisolated` is required: this is awaited from a `Task` while
    // `syncBatchBlocking` blocks the actor's executor on a semaphore during the
    // auto-sync path. If it were actor-isolated, that Task could never enter the
    // (blocked) actor and the semaphore would never be signaled — a deadlock
    // that silently killed custom-body auto-sync on iOS. It touches no actor
    // state, so isolation is unnecessary.
    nonisolated func executeFallbackHttpSync(coreHttp: HttpConfig, customBody: String, interceptor: DartSyncInterceptor?) async -> FallbackSyncResult {
        var currentHeaders = coreHttp.headers
        let maxRetries = Int(coreHttp.maxRetries)
        
        var lastStatus = 0
        var lastResponse = "Unknown error"
        
        for attempt in 0...maxRetries {
            guard let urlStr = coreHttp.url, let url = URL(string: urlStr) else { 
                return FallbackSyncResult(success: false, status: 0, responseText: "Invalid URL") 
            }
            
            var request = URLRequest(url: url)
            request.httpMethod = coreHttp.method == 1 ? "PUT" : "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.timeoutInterval = 15.0
            
            for (k, v) in currentHeaders {
                request.setValue(v, forHTTPHeaderField: k)
            }
            
            request.httpBody = customBody.data(using: .utf8)
            
            do {
                let (data, response) = try await URLSession.shared.data(for: request)
                if let httpResponse = response as? HTTPURLResponse {
                    let status = httpResponse.statusCode
                    lastStatus = status
                    lastResponse = String(data: data, encoding: .utf8) ?? ""
                    
                    if (200...299).contains(status) {
                        return FallbackSyncResult(success: true, status: status, responseText: lastResponse)
                    } else if status == 401, let interceptor = interceptor {
                        if interceptor.requestTokenRefresh() {
                            if let newConfig = TraceletSdk.shared.rustEngineState?.getConfig().http {
                                currentHeaders = newConfig.headers
                            }
                            continue
                        } else {
                            return FallbackSyncResult(success: false, status: status, responseText: lastResponse)
                        }
                    } else if (400...499).contains(status) && status != 408 && status != 429 {
                        // Client error, no point in retrying (except timeout/rate limits)
                        return FallbackSyncResult(success: false, status: status, responseText: lastResponse)
                    }
                }
            } catch {
                TraceletSdk.shared.logger.debug("Fallback sync error: \(error)")
                lastResponse = error.localizedDescription
                lastStatus = 0
            }
            
            if attempt < maxRetries {
                try? await Task.sleep(nanoseconds: UInt64(1_000_000_000 * (attempt + 1)))
            }
        }
        return FallbackSyncResult(success: false, status: lastStatus, responseText: lastResponse)
    }
}

class TraceletSyncSink: LocationDataSink, SyncProvider {
    let coordinator = SyncCoordinator()
    
    @discardableResult
    func insertLocation(_ location: [String: Any]) -> String {
        Task {
            await coordinator.scheduleSync(sink: self)
        }
        return ""
    }

    // Cancel the in-flight debounce on stop() so background sync halts
    // immediately instead of firing ~autoSyncDelay later (#213).
    func cancelPendingSync() {
        Task { await coordinator.cancel() }
    }
    
    func syncBatchBlocking(config: HttpConfig, records: [DbLocationRecord]) throws -> UInt32 {
        let syncRecords: [tracelet_sync.SyncLocationRecord] = records.map { r in
            tracelet_sync.SyncLocationRecord(
                id: r.id,
                uuid: r.uuid,
                timestamp: r.timestamp,
                latitude: r.latitude,
                longitude: r.longitude,
                accuracy: r.accuracy,
                speed: r.speed,
                heading: r.heading,
                altitude: r.altitude,
                isMock: r.isMock,
                isMoving: r.isMoving,
                activity: r.activity,
                event: r.eventType,
                routeContext: r.routeContext,
                address: r.address  // #212: carry reverse-geocoded address into the default payload
            )
        }
        
        // INTERCEPTOR LOGIC
        if let interceptor = TraceletSdk.shared.dartSyncInterceptor {
            if interceptor.requestFreshHeaders() {
                TraceletSdk.shared.logger.debug("Headers refreshed in syncBatchBlocking")
            }
            let updatedHttp = TraceletSdk.shared.rustEngineState?.getConfig().http ?? config
            
            // Issue #126: emit the SAME nested schema as onLocation/getLocations
            // (nested coords/activity/battery + route_context) so the Dart
            // custom-body builder receives a consistent shape instead of a flat
            // map with a raw String activity.
            let syncRecordsMap: [[String: Any]] = records.map {
                TraceletSdk.shared.mapRecordToLocation($0)
            }
            
            let customBody = interceptor.requestSyncBody(locations: syncRecordsMap)
            if customBody == nil {
                // A builder is registered but it timed out or threw.
                // requestSyncBody returns the sentinel (not nil) when no builder
                // exists, so nil here unambiguously means failure: abort the sync
                // rather than posting an error object or the default payload.
                TraceletSdk.shared.logger.error("Custom sync body failed to build; aborting sync.")
                TraceletSdk.shared.getEventSender().sendHttp([
                    "success": false,
                    "status": 0,
                    "responseText": "custom sync body failed to build",
                    "isRetry": false,
                    "retryCount": 0
                ])
                return 0
            }
            if let body = customBody, body != traceletNoSyncBodyBuilderSentinel {
                TraceletSdk.shared.logger.debug("customBody from interceptor (syncBatchBlocking): \(body)")
                let sem = DispatchSemaphore(value: 0)
                var fallbackResult: SyncCoordinator.FallbackSyncResult? = nil

                Task {
                    fallbackResult = await coordinator.executeFallbackHttpSync(coreHttp: updatedHttp, customBody: body, interceptor: interceptor)
                    sem.signal()
                }

                sem.wait()

                if let result = fallbackResult, result.success {
                    // #214 dedup: the custom builder may have included the
                    // telematics we exposed; the POST succeeded, so mark exactly
                    // those synced so they aren't re-sent.
                    TraceletSdk.shared.markExposedTelematicsSynced()
                    TraceletSdk.shared.getEventSender().sendHttp([
                        "success": true,
                        "status": result.status,
                        "responseText": result.responseText,
                        "isRetry": false,
                        "retryCount": 0
                    ])
                    return UInt32(records.count)
                } else {
                    let result = fallbackResult ?? SyncCoordinator.FallbackSyncResult(success: false, status: 0, responseText: "Unknown error")
                    TraceletSdk.shared.logger.debug("Custom body sync failed with status \(result.status)")
                    TraceletSdk.shared.getEventSender().sendHttp([
                        "success": false,
                        "status": result.status,
                        "responseText": result.responseText,
                        "isRetry": false,
                        "retryCount": 0
                    ])
                    return 0
                }
            }
            // sentinel → no builder → fall through to the default sync below.
        }

        let syncHttp = tracelet_sync.SyncHttpConfig(
            url: config.url,
            method: config.method,
            headers: config.headers,
            batchSync: config.batchSync,
            maxBatchSize: config.maxBatchSize,
            autoSync: config.autoSync,
            maxRetries: config.maxRetries,
            retryBackoffBase: config.retryBackoffBase,
            retryBackoffCap: config.retryBackoffCap,
            sslPinningCertificates: config.sslPinningCertificates,
            sslPinningFingerprints: config.sslPinningFingerprints,
            httpRootProperty: config.httpRootProperty,
            params: config.params,
            extras: config.extras,
            disableAutoSyncOnCellular: config.disableAutoSyncOnCellular,
            enableDeltaCompression: config.enableDeltaCompression,
            deltaCoordinatePrecision: config.deltaCoordinatePrecision,
            locationsOrderDirection: config.locationsOrderDirection
        )
        
        let syncManager = SyncManager()
        do {
            let count = try syncManager.syncBatchBlocking(config: syncHttp, records: syncRecords)
            if count > 0 {
                TraceletSdk.shared.getEventSender().sendHttp([
                    "success": true,
                    "status": 200,
                    "responseText": "Synced \(count) locations",
                    "isRetry": false,
                    "retryCount": 0
                ])
            } else {
                TraceletSdk.shared.getEventSender().sendHttp([
                    "success": false,
                    "status": 0,
                    "responseText": "Sync failed",
                    "isRetry": false,
                    "retryCount": 0
                ])
            }
            return count
        } catch {
            TraceletSdk.shared.getEventSender().sendHttp([
                "success": false,
                "status": 0,
                "responseText": error.localizedDescription,
                "isRetry": false,
                "retryCount": 0
            ])
            throw error
        }
    }
}

@objc(TraceletSyncPlugin)
public class TraceletSyncPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "tracelet_sync", binaryMessenger: registrar.messenger())
    let instance = TraceletSyncPlugin()
    
    let sink = TraceletSyncSink()
    TraceletSdk.shared.locationEngine.registerSink(sink)
    TraceletSdk.shared.syncProvider = sink
    print("TraceletSync: Native iOS sink registered!")

    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    if call.method == "initialize" {
      result(true)
    } else {
      result(FlutterMethodNotImplemented)
    }
  }
}
